core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala

```diff
@@ -180,7 +183,42 @@ private[spark] abstract class BasePythonRunner[IN, OUT](
         dataOut.writeInt(partitionIndex)
         // Python version of driver
         PythonRDD.writeUTF(pythonVer, dataOut)
+        // Init a GatewayServer to port current BarrierTaskContext to Python side.
+        val isBarrier = context.isInstanceOf[BarrierTaskContext]
+        val secret = if (isBarrier) {
+          Utils.createSecret(env.conf)
+        } else {
+          ""
+        }
+        val gatewayServer: Option[GatewayServer] = if (isBarrier) {
+          Some(new GatewayServer.GatewayServerBuilder()
+            .entryPoint(context.asInstanceOf[BarrierTaskContext])
+            .authToken(secret)
+            .javaPort(0)
+            .callbackClient(GatewayServer.DEFAULT_PYTHON_PORT, GatewayServer.defaultAddress(),
+              secret)
+            .build())
```

Comment:
We have to port `BarrierTaskContext` from java to python side, otherwise there is no way to call `BarrierTaskContext.barrier()` from python side. Thus, of course, the JavaGateway is only initiated when the context is a `BarrierTaskContext`.

Reply:
Yea, I read and understood if this is only initialised when the context is a `BarrierTaskContext` but this is super weird we start another Java gateway here. If it's a hard requirement, then I suspect the design issue. Should this be targeted to 2.4.0, @mengxr?

Reply:
If this should necessarily target 2.4.0, don't block by me since it's a new feature and probably we could consider another approach later but if we can avoid, I would suggest to avoid for now. 

Let me try to track the design doc and changes about this. I think I need more time to check why it happened like this and if there's another way.

Reply:
@HyukjinKwon Could you elaborate your concerns? Is it because resource usage or security?

Reply:
Mainly the reason is about resource usage, unusual access pattern via Py4J at worker side, and the possibility of allowing JVM access within Python worker.

It pretty much looks an overkill to launch a Java gateway to allow access to call a function assuming from https://github.com/apache/spark/pull/22085#discussion_r209490553. This pattern sounds pretty unusual - such cases, we usually send the data manually and read it in Python side, for instance `TaskContext`. Now, it opens a gateway for each worker if I am not mistaken.

I was thinking if we can avoid this. Can you elaborate why this is required and necessary? I haven't got enough time to look into this so was thinking about taking a look on this weekends.

This also now opens a possibility for an JVM access from worker side via `BarrierTaskContext`. For instance, I believe we can hack and access to JVM inside of UDFs.

Reply:
BTW, this is a design change. We should probably update https://cwiki.apache.org/confluence/display/SPARK/PySpark+Internals too .. is this really required to open a gate there?

Reply:
The major issue here is that we want to make the `barrier()` call blocking, the task shall wait until timeout or succeeded, do we have other ways to achieve this goal other than current approach here?

Reply:
Not yet. So I asked to hold it back for now since another gateway here looks the last choice, and was wondering if we can avoid to target 2.4.0. If this blocks, please go ahead. Will check it later on this weekends.